xcdat/climatology.py

```diff
@@ -0,0 +1,191 @@
+        month_length.groupby(time_period) / month_length.groupby(time_period).sum()
+    )
+    _validate_weights(ds, weights, time_period)
+    return weights
+
+
+def _validate_weights(ds: xr.Dataset, weights: xr.DataArray, time_period: TimePeriod):
+    """Validate that the sum of the weights for a dataset equal 1.0.
+
+    :param ds: A Dataset object
+    :type ds: xr.Dataset
+    :param weights: The calculated weights based on a period of time
+    :type weights: xr.DataArray
+    :param time_period: The period of time to group data by
+    :type time_period: TimePeriod
+    """
+    expected_count = {
+        "time.month": 12,
+        "time.season": 4,
+        "time.year": np.unique(ds.time.dt.year.data).size,
+    }
+
+    expected = np.ones(expected_count[time_period])
+    np.testing.assert_allclose(weights.groupby(time_period).sum().values, expected)
+
+
+def departure(
+    ds_base: xr.Dataset,
+    ds_climatology: xr.Dataset,
+) -> xr.Dataset:
+    """Calculate departures for a given climatology.
+
+    :param ds: The base dataset
+    :type ds: xr.Dataset
+    :param ds_climatology: The climatology dataset
+    :type ds_climatology: xr.Dataset
+    :raises KeyError: When the incorrect climatology dataset is passed
+    :return: A climatology departure dataset
+    :rtype: xr.Dataset
+    """
+    time_period = _extract_period(ds_climatology)
+    weights = None
+
+    if ds_climatology.attrs["is_weighted"] is True:
+        weights = _calculate_weights(ds_base, time_period)  # type: ignore
+
+    ds_departure = _group_data(ds_base.copy(), time_period, "departure", weights)  # type: ignore
+
+    for key in ds_departure.data_vars.keys():
+        if "_bnds" not in str(key):
+            ds_departure[key] = ds_departure[key] - ds_climatology[key]
+
+    return ds_departure
+
+
+def _group_data(
+    ds: xr.Dataset,
+    time_period: TimePeriod,
+    calculation_type: Literal["climatology", "departure"],
+    weights: Optional[xr.DataArray] = None,
+) -> xr.Dataset:
+    """Groups data variables to get the average for a time period.
+
+    Grouping only applies to non-bounds variables, which perserves any
+    existing bounds variables. The "time" dimension and its associated
+    coordinates and variables are also preserved since individual data
+    variables are grouped, rather than the entire dataset.
+
+    :param ds: A Dataset object
+    :type ds: xr.Dataset
+    :param time_period: The period of time to group data by
+    :type time_period: TimePeriod
+    :param weights: The calculated weights based on period of time, defaults to None
+    :type weights: Optional[xr.DataArray], optional
+    :return: The calculated climatology
+    :rtype: xr.Dataset
+    """
+    ds_grouped = ds.copy()
+    ds_grouped.attrs.update({"calculation_type": calculation_type})
+    if weights is not None:
+        ds_grouped.attrs.update({"is_weighted": True})
+    else:
+        ds_grouped.attrs.update({"is_weighted": False})
```

Comment:
@pochedls @jasonb5 @chengzhuzhang @lee1043  We can add attributes to DataArrays and Datasets by using their `.attrs` dict. These attributes can be referenced as if they were native to xarray.

```
# Functional programming with dataset attributes
# ==================================
ds = xr.Dataset()

# The `is_weighted` dataset attribute is set to True
ds_climo = climatology(ds, "season", weighted=True)
ds_climo.is_weighted # True

ds_departure = departure(ds, ds_climo)

# Class-based accessors with class attributes
# ================================
ds = xr.Dataset()

# The self.is_weighted class attribute is set to True
ds_climo = ds.climatology.climatology("season", weighted=True)
ds_climo.is_weighted # True

ds_departure = ds.climatology.depature(ds_climo)
```

If the goal is API flexibility with custom attributes (e.g., `is_weighted`) and ease of implementation/usage, this is a good alternative over implementing class-based accessors.